### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by daria-sieroshtan**

### CODE REVIEW

File: src/Controller/CategoryController.php [hunk 1]
```diff
@@ -40,8 +41,49 @@ public function listCategory(Request $request)
...
-        $this->addNoticeFlash('Functionality is not available yet.');
-        return $this->redirectToRoute('list_categories');
+        $category = new Category;
+        $form = $this->createForm(CategoryType::class, $category);
+        $form->handleRequest($request);
+        
+        if ($form->isSubmitted() && $form->isValid()) {
+            $category = $form->getData();
+        
+            $entityManager = $this->getDoctrine()->getManager();
```
Comment (arodiss) [inline src/Controller/CategoryController.php]: @daria-sieroshtan 
This is not a mistake, but I suggest to create helper method `getEm()` in base controller. It is needed very, very often

File: src/Controller/CategoryController.php [hunk 2]
```diff
@@ -40,8 +41,49 @@ public function listCategory(Request $request)
     */
     public function addCategory(Request $request)
     {
-        $this->addNoticeFlash('Functionality is not available yet.');
-        return $this->redirectToRoute('list_categories');
+        $category = new Category;
+        $form = $this->createForm(CategoryType::class, $category);
```
Comment (arodiss) [inline src/Controller/CategoryController.php]: @daria-sieroshtan 
This is not a mistake, but you can do it inline:
```PHP
$form = $this->createForm(CategoryType::class, new Category());
```

File: src/Controller/CategoryController.php [hunk 3]
```diff
@@ -40,8 +41,49 @@ public function listCategory(Request $request)
...
+        }
+        
+        return $this->render('category/category_add.html.twig', array(
+            'form' => $form->createView(),
+        ));
+    }
+    
+    /**
+    * @Route("/category/edit/{id}", name="edit_category", requirements={"id"="\d+"})
+    * @Route("/category/edit/{slug}", name="edit_category_by_slug")
```
Comment (arodiss) [inline src/Controller/CategoryController.php]: @daria-sieroshtan 
This is not a mistake, but generally one is enough

File: src/Controller/CategoryController.php [hunk 4]
```diff
@@ -40,8 +41,49 @@ public function listCategory(Request $request)
...
+        $form->handleRequest($request);
+        
+        if ($form->isSubmitted() && $form->isValid()) {
+            $category = $form->getData();
+        
+            $entityManager = $this->getDoctrine()->getManager();
+            $entityManager->persist($category);
+            $entityManager->flush();
+
+            $this->addNoticeFlash('Successfully edited category "' . $category->getName() . '"');
```
Comment (arodiss) [inline src/Controller/CategoryController.php]: @daria-sieroshtan 
`sprintf()`
It is going to be required when we add translations

File: src/Controller/JoyController.php [hunk 5]
```diff
@@ -39,8 +42,49 @@ public function listJoy(Request $request)
...
+        
+            $entityManager = $this->getDoctrine()->getManager();
+            $entityManager->persist($joy);
+            $entityManager->flush();
+
+            $this->addNoticeFlash('Successfully edited joy "' . $joy->getName() . '"');
+            return $this->redirectToRoute('list_joys');
+        }
+        
+        return $this->render('joy/joy_edit.html.twig', array(
```
Comment (arodiss) [inline src/Controller/JoyController.php]: @daria-sieroshtan 
This is not a mistake, but usually people use `@Template` annotation

File: src/Form/CategoryType.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+use Symfony\Component\Form\Extension\Core\Type\SubmitType;
+
+class CategoryType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('name', TextType::class)
+            ->add('isPhysical')
+            ->add('isCultural')
```
Comment (arodiss) [inline src/Form/CategoryType.php]: @daria-sieroshtan 
Why don't you put a type here?

File: src/Form/JoyType.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            ->add('name', TextType::class)
+            ->add('description', TextareaType::class)
+            ->add('date', DateType::class)
+            ->add('place', EntityType::class, array(
+            'class' => Place::class,
+            'choice_label' => 'name'))
+            ->add('categories', EntityType::class, array(
+            'class' => Category::class,
+            'choice_label' => 'name',
+            'multiple' => true))
```
Comment (arodiss) [inline src/Form/JoyType.php]: @daria-sieroshtan 
Misindent.
Should be:
```PHP
$builder
            ->add('place', EntityType::class, array(
                'class' => Place::class,
                'choice_label' => 'name'
            ))
            ->add('categories', EntityType::class, array(
                'class' => Category::class,
                'choice_label' => 'name',
                'multiple' => true
            ))
```
Also people generally prefer short array syntax, ie `[1, 2]` over`array(1, 2)`

File: src/Form/JoyType.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            ->add('save', SubmitType::class)
```
Comment (arodiss) [inline src/Form/JoyType.php]: @daria-sieroshtan 
Best practises suggest other way to do it

File: templates/category/category_add.html.twig [hunk 9]
```diff
@@ -0,0 +1,9 @@
+{% extends 'base.html.twig' %}
+
+{%block heading %}Create new category{% endblock %}
+
+{% block content %}
+    {{ form_start(form) }}
+    {{ form_widget(form) }}
+    {{ form_end(form) }}
```
Comment (arodiss) [inline templates/category/category_add.html.twig]: @daria-sieroshtan 
If you don't add anything except this, you can simply do `{{ form(form) }}`
Also please notice that your form templates (**4 of them!**) are identical except page title. Don't copy-paste, use template inheritance

File: src/Controller/BaseController.php [hunk 10]
```diff
@@ -19,4 +19,13 @@ public function addNoticeFlash($message)
     {
         $this->addFlash('notice', $message);
     }
+    
+    /**
+     *
+     * @return entityManager
```
Comment (arodiss) [inline src/Controller/BaseController.php]: @daria-sieroshtan 
MB `EntityManager`? Also for this annotation to work, class `EntityManager` should be imported in `BaseController`

File: src/Controller/CategoryController.php [hunk 11]
```diff
@@ -25,34 +27,68 @@ public function listCategory(Request $request)
...
+        $category = new Category;
+        $form = $this->createForm(CategoryType::class, $category);
+        $form->handleRequest($request);
+        
+        if ($form->isSubmitted() && $form->isValid()) {
+            $category = $form->getData();
+        
+            $entityManager = $this->getEm();
+            $entityManager->persist($category);
+            $entityManager->flush();
```
Comment (arodiss) [inline src/Controller/CategoryController.php]: @daria-sieroshtan 
The goal of helper method is to do
```PHP
$this->getEm()->persist($entity);
$this->getEm()->flush();
```

File: src/Form/CategoryType.php (hunk 6 above)
Comment (arodiss) [reply src/Form/CategoryType.php]: ?

File: src/Form/JoyType.php [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        $builder
+            ->add('name', TextType::class)
+            ->add('description', TextareaType::class)
+            ->add('date', DateType::class)
+            ->add(
+                'place',
+                EntityType::class,
+                [
+                'class' => Place::class,
```
Comment (arodiss) [inline src/Form/JoyType.php]: @daria-sieroshtan 
Inner should be indented more than outer

File: src/Form/CategoryType.php (hunk 6 above)
Comment (daria-sieroshtan) [reply src/Form/CategoryType.php]: There is no benefit in it except of harm

Comment (arodiss) [reply src/Form/CategoryType.php]: @daria-sieroshtan 
As a great man said: *explicit is better than implicit*

**Pull request merged**